qa/tasks/ceph_salt.py

```diff
@@ -0,0 +1,274 @@
+        if not isinstance(ceph_salt_ctx['log_anchor'], str):
+            self.log.warning(
+                "log_anchor was set to non-string value ->{}<-, "
+                "changing to empty string"
+                .format(ceph_salt_ctx['log_anchor'])
+                )
+            ceph_salt_ctx['log_anchor'] = ''
+        ceph_salt_ctx['deploy'] = self.config.get('deploy', True)
+        ceph_salt_ctx['quiet_salt'] = self.config.get('quiet_salt', True)
+        ceph_salt_ctx['salt_manager_instance'] = SaltManager(self.ctx)
+        ceph_salt_ctx['master_remote'] = (
+                ceph_salt_ctx['salt_manager_instance'].master_remote
+                )
+        ceph_salt_ctx['repo'] = self.config.get('repo', None)
+        ceph_salt_ctx['branch'] = self.config.get('branch', None)
+
+    def _add_CA_repo(self):
+        '''
+        Adding `SUSE Internal CA Certificate` that's needed to install the container image
+        '''
+        self.ctx.cluster.run(args=[ 'sudo', 'sh', '-c', 
+                                    'sudo zypper -n addrepo --refresh --no-gpgcheck'
+                                    ' http://download.suse.de/ibs/SUSE:/CA/SLE_15_SP2/ '
+                                    '\'SUSE Internal CA Certificate\'; '
+                                    'zypper -n in ca-certificates-suse'])
+
+    def _rm_localhost_from_etc_hosts(self):
+        '''
+        Teuthology by default puts in '/etc/hosts' a line with the hostnames and localhost ip. This
+        causes salt to populate 'fqdn_ip4' with 127.0.0.1 and ceph-salt is breaking since it
+        has a exception for localhost resolution. So this function Removes localhost entries from
+        '/etc/hosts' in all hosts in order to use DNS server for names resolution.
+        '''
+        self.log.info(anchored("Removing localhost resolution from /etc/hosts"))
+        self.ctx.cluster.run(args=[ 'sudo', 'sh', '-c', 
+                                    'sed -i \"/$(hostname)/d\" /etc/hosts' ])
+        self.log.info("Removed localhost resolution")
+        self.ctx.cluster.run(args=[ 'sudo', 'sh', '-c',
+                                    'cat /etc/hosts' ])
+        self.ctx.cluster.run(args=[ 'sudo', 'sh', '-c',
+                                    'chattr +i /etc/hosts'])
+
+    def _ceph_salt_config(self):
+        '''
+        This function populates ceph-salt config according to the configuration on the yaml files
+        on the suite regarding node roles, chrony server, dashboard credentials etc and then runs
+        the cluster deployment
+        '''
+        ceph_salt_roles = {"mon": "Mon", "mgr": "Mgr"}
+        for host, roles in self.remote_lookup_table.items():
+            self.master_remote.sh("sudo ceph-salt config /Cluster/Minions add {}".format(host))
+            for role in roles:
+                role = role.split('.')[0]
+                if role in ceph_salt_roles.keys():
```

Comment:
You do not need to get use `.keys()` to check if a dictionary has it in python, here is more pythonish style:
```
if role in ceph_salt_roles:
   cluster_role = '/Cluster/Roles/' + ceph_salt_roles[role]
   self.master_remote.sh("sudo ceph-salt config {} add {}".format(cluster_role, host))
```

Reply:
OK I aggree with you on the `.keys` part but I think it's more readable to have the whole command instead of having `cluster_role`

Reply:
for python dev it should be readable ;-)